plugins/colorbutton/plugin.js

```diff
@@ -370,6 +414,152 @@ CKEDITOR.plugins.add( 'colorbutton', {
+			CKEDITOR.tools.array.forEach( contentColorsRow.find( 'td' ).toArray(), function( node ) {
+				node.remove();
+			}, 0, 1 );
+
+			if ( !colorSpans.length ) {
+				contentColorsLabel.hide();
+				return;
+			}
+
+			CKEDITOR.tools.array.forEach( colorSpans, function( span ) {
+				// colorSpans may contain spans with background color when they are not necessary, so here they are being filtered out.
+				var spanColor = span.getStyle( cssAttribute );
+
+				if ( spanColor ) {
+					colors.push( normalizeColor( spanColor ).toUpperCase() );
+				}
+			} );
+
+			if ( !colors.length ) {
+				contentColorsLabel.hide();
+				return;
+			}
+
+			CKEDITOR.tools.array.forEach( colors, function( color ) {
+				if ( color in colorOccurrences ) {
+					colorOccurrences[ color ]++;
+				} else {
+					colorOccurrences[ color ] = 1;
+				}
+			} );
+
+			for ( var color in colorOccurrences ) {
+				sortedColors.push( [ color, colorOccurrences[ color ] ] );
+			}
+			sortedColors.sort( function( a, b ) {
+				return b[ 1 ] - a[ 1 ];
+			} );
+
+			// Trim the color array to the row size.
+			if ( sortedColors.length > colorsPerRow ) {
+				sortedColors.splice( colorsPerRow - 1, sortedColors.length - colorsPerRow );
+			}
+
+			CKEDITOR.tools.array.forEach( sortedColors, function( color ) {
+				color[ 2 ] = editor.lang.colorbutton.colors[ color[ 0 ] ] || color[ 0 ];
+			} );
+
+			CKEDITOR.tools.array.forEach( sortedColors, function( color ) {
+				// Unfortunately CKEDITOR.dom.element.createFromHtml() doesn't work for table elements,
+				// so table cell has to be created separately.
+				var colorTile = new CKEDITOR.dom.element( 'td' );
+
+				colorTile.setHtml( generateTileHtml( {
+					colorLabel: color[ 2 ],
+					clickFn: clickFn,
+					colorCode: color[ 0 ],
+					type: type,
+					position: tilePosition++,
+					setSize: sortedColors.length
+				} ) );
+
+				contentColorsRow.append( colorTile );
+			} );
+
+			contentColorsLabel.show();
+		}
+
+		function addCustomColorToPanel( options ) {
+			// This function is called when a color from colordialog is chosen.
+			var colorHexCode = options.colorHexCode,
+				colorLabel = editor.lang.colorbutton.colors[ colorHexCode ] || colorHexCode,
+				clickFn = options.clickFn,
+				type = options.type,
+				colorsPerRow = options.colorsPerRow,
+				customColorsRow = options.customColorsRow,
+				customColorsLabel = options.customColorsLabel,
+				chosenColorTile = customColorsRow.findOne( '[data-value="' + colorHexCode + '"]' ),
+				tilesNumber = customColorsRow.getChildCount(),
+				existingTiles;
+
+			if ( chosenColorTile ) {
+				// If the same color is chosen again, find the old tile and move it to the beginning
+				// instead of creating a new one.
+				customColorsRow.append( chosenColorTile.getParent(), true );
+			} else {
+				var colorTile = new CKEDITOR.dom.element( 'td' );
+
+				tilesNumber++;
```

Comment:
It looks like there is some loop over this variable. But if I see correctly, it's just incremented to match correct count? If so, maybe it could be done right at the definition to avoid misunderstanding?

Reply:
It would have to be calculated twice then, as it should be incremented when new color is picked, while for existing color it's value won't change. But according to your comment \[here\](https://github.com/ckeditor/ckeditor4/pull/3784#discussion\_r367987985) I changed `++` to `+= 1`.